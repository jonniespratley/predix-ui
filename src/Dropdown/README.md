# px-dropdown
The `<Dropdown/>` component has two primary use cases. First, as a form control, dropdown lists are ideal for a user to choose from a preconfigured list of options â€“ ideally, between 5 and 15. If there are fewer than 5 options, consider using a radio button. For very large lists of options, consider adding typeahead functionality in addition to, or instead of, a dropdown. The second primary use case for a dropdown list is as a filter or navigation item, where invoking an option from the dropdown menu results in an action of some kind.



## Usage


```react
state:
  selectedItem: null
---
const { Dropdown } = PxReact;
const items = [
  {"key":"1","val":"iPhone"},
  {"key":"2","val":"Android"},
  {"key":"3","val":"Blackberry"},
  {"key":"4","val":"Windows Phone"},
  {"key":"5","val":"Flip Phone","disabled":true}
];
<div>
  <Dropdown 
    onChange={(e) => setState({selectedItem: e})}
    items={items}
    sortMode="key" 
    buttonStyle="default" 
    displayValue="Select">
  </Dropdown>
  <pre>selectedItem: {JSON.stringify(state.selectedItem, null, 2)}</pre>
</div>
```

```react
state:
  selectedItem: null
---
const { Dropdown } = PxReact;
const items = [
  {"key":"1","val":"iPhone"},
  {"key":"2","val":"Android"},
  {"key":"3","val":"Blackberry"},
  {"key":"4","val":"Windows Phone"},
  {"key":"5","val":"Flip Phone","disabled":true}
];
<div>
  <Dropdown 
    onChange={(e) => setState({selectedItem: e})}
    items={items}
    sortMode="key" 
    buttonStyle="default" 
    displayValue="Select" 
    searchMode
    multi >
  </Dropdown>
  <pre>selectedItem: {JSON.stringify(state.selectedItem, null, 2)}</pre>
</div>
```


## Properties

```table
span: 6
rows:
  - Name: buttonStyle
    Type: String
    Description: The style for the button.
  - Name: items
    Type: Array
    Description: Array of dropdown items.
  - Name: displayValue
    Type: String
    Description: The value to display.
```


### Styling

```table
span: 6
rows:
  - Property: --my-prop
    Default: null
    Description: This is the style prop.
```
