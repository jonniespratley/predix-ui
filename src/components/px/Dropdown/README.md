# px-dropdown
The `<dropdown/>` component enables you to ...



## Usage

```react
const { Dropdown } = PxReact;
<Dropdown>
  This is the children
</Dropdown>
```


## Properties

```table
span: 6
rows:
  - Name: children
    Type: node
    Description: The children content.
```


### Styling

```table
span: 6
rows:
  - Property: --my-prop
    Default: null
    Description: This is the style prop.
```
