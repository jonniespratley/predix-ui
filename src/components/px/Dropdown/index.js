import React from 'react';
import classnames from 'classnames';
//import stylesheet from './px-dropdown.scss';
import BaseComponent from '../BaseComponent';
import Icon from '../IconSet/px-icon';
import styled, {css} from 'styled-components';
import Button from '../Button';


const SearchInput = styled.input`
  height: 2em;
  width: 100%;
  border: none;
  border-radius: 0;
  border-bottom: 1px solid var(--px-input-border-color--outer,#889aa5);
  padding: 0 .33333rem;
  color: var(--px-input-text-color,#2c404c);
  &:focus{
    border-color: var(--px-input-border-color--outer--focused,#007acc);
    background-color: var(--px-input-background-color--focused,#eefbff);
    color: var(--px-input-text-color--focused,inherit);
  }
`;

const DropdownLabel = styled.div`
  overflow: hidden;
  flex-grow: 1;
  text-overflow: ellipsis;
  text-align: left;
  
`;
DropdownLabel.displayName = 'DropdownLabel';

const DropdownContent = styled.div`
  z-index: 10;
  position: absolute;
  background-color: var(--px-dropdown-bg-color,#fff);
  border: 1px solid var(--px-dropdown-border-color,gray);
  max-height: var(--px-dropdown-max-height,300px);
  max-width: var(--px-dropdown-max-width,400px);
  min-width: var(--px-dropdown-min-width);
  min-height: var(--px-dropdown-min-height);
  display: flex;
  flex-direction: column;
  display: none;
  ${props => props.opened && css`
    display: block;
  `}
  
`;
DropdownContent.displayName = 'DropdownContent';

const DropdownOption = styled.div`
    padding-left: var(--px-dropdown-item-padding,.66667rem);
    padding-right: var(--px-dropdown-item-padding,.66667rem);
    padding-top: var(--px-dropdown-item-padding);
    padding-bottom: var(--px-dropdown-item-padding);
    height: var(--px-dropdown-item-height,2rem);
    cursor: pointer;
    display: flex;
    align-items: center;
    &:hover {
      background-color: var(--px-dropdown-bg-color--hover,#d3d3d3);
    }
    &.selected,
    &:active {
      background-color: var(--px-dropdown-bg-color--selected,#00f);
      color: var(--px-dropdown-text-color--selected,#fff);
    }
    ${props => props.disabled && css`
      color: var(--px-dropdown-text-color--disabled,gray);
      background-color: var(--px-dropdown-bg-color--disabled,transparent);
      cursor: not-allowed;
      pointer-events: none;
    `}
    
    ${props => props.selected && css`
      background-color: var(--px-dropdown-bg-color--selected,#00f);
      color: var(--px-dropdown-text-color--selected,#fff);
      &:hover {
        background-color: var(--px-dropdown-bg-color--selected,#00f);
        color: var(--px-dropdown-text-color--selected,#fff);
      }
    `}
`;
DropdownOption.displayName = 'DropdownOption';

const DropdownTrigger = Button.extend`
  max-width: inherit;
  width: inherit;
  display: flex;
  justify-content: center;
  align-items: center;
  padding-left: .66667rem;
  padding-right: .66667rem;
  outline: 0;
  font-weight: var(--px-dropdown-font-weight,normal);
  height: auto;
`;
DropdownTrigger.displayName = 'DropdownTrigger';



/**
 * px-example-component component
 */
class Dropdown extends BaseComponent {
  constructor(props){
    super(props, {name: 'Dropdown'});
    this.state = {
      selected: props.selected || null,
      selectedItems: props.selectedItems || null,
      selectedValues: props.selectedValues || null,
      opened: props.opened || false,
      items: props.items || null
    };
  }
  _handleClick(e){
    this.setState({opened: !this.state.opened});
  }
  _clearSelected(){
    let items = this.state.items;
    items.map(item => item.selected = false);
    this.setState({items});
  }
  _handleChange(item, index, e){
    console.log(item, index);
    if(item.disabled){
      return;
    }
    
    this.setState({
      selectedItem: item
    });
    
    if(!this.props.multi){
      this._clearSelected();
      this._handleClick();
    } else {
      console.log('Push item to selectedValues');  
    }
    let items = this.state.items[index].selected = !this.state.items[index].selected;
    

    if (this.props.onChange) {
      this.props.onChange(item, index, e);
    }
  }
	render() {
		const {
			label = 'px-example-component',
      selectBy,
      disabled,
      disableClear,
      hover,
      icon,
      triggerHeight,
      focusedItem,
      hideChevron,
      buttonStyle,
      boundTarget,
      preventCloseOnOutsideClick,
      displayValue,
      multi,
			children
    } = this.props;
    const triggerStyle = {
      height: triggerHeight 
    };
    const {opened, items, selectedItems, selectedValues, selected} = this.state;
		const baseClasses = classnames('px-dropdown');
		return (
			<div className={baseClasses}>
        <DropdownTrigger {...buttonStyle} disabled={disabled} onClick={this._handleClick.bind(this)} style={triggerStyle}>
          <DropdownLabel>{displayValue}</DropdownLabel>
          {!hideChevron && <Icon icon={icon} size={16} />}
        </DropdownTrigger>
        <DropdownContent opened={opened}>
          {items && items.map((item, index) => (
            <DropdownOption 
              disabled={item.disabled}
              selected={item.selected}
              onClick={this._handleChange.bind(this, item, index)}
              key={item.key}>{item.val}
            </DropdownOption>)
          )}
        </DropdownContent>      
			</div>
		);
	}
}

Dropdown.defaultProps = {
  icon: 'px-utl:chevron',
  showCaret: false,
  hideChevron: true
};

Dropdown.displayName = 'Dropdown';

export default Dropdown;