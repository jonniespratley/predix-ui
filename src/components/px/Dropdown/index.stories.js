import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withKnobs, text, array, boolean, select } from '@storybook/addon-knobs';
import { withReadme } from 'storybook-readme';

import Dropdown from './';
import README from './README.md';
import Icon from '../IconSet/Icon';

const buttonStyles = [
  'bare',
  'bare-primary',
  'default',
  'tertiary',
  'icon'
];

const items = [
  { key: '1', val: 'iPhone' },
  { key: '2', val: 'Android' },
  { key: '3', val: 'Blackberry' },
  { key: '4', val: 'Windows Phone' },
  { key: '5', val: 'Flip Phone', disabled: true }
];

storiesOf('Components / Dropdown', module)
  .addDecorator(withKnobs)
  .addDecorator(withReadme(README))
  .addWithJSX('default', () => (
    <Dropdown
      onChange={action('onChange')}
      items={array('items', items)}
      sortMode="key"
      opened={boolean('opened', false)}
      disabled={boolean('disabled', false)}
      disableClear={boolean('disableClear', false)}
      hideChevron={boolean('hideChevron', false)}
      buttonStyle={select('buttonStyle', buttonStyles, 'default')}
      displayValue={text('displayValue', 'Select')}
    />
  ))
  .addWithJSX('with custom trigger', () => (
    <Dropdown
      onChange={action('onChange')}
      items={array('items', items)}
      sortMode="key"
      disabled={boolean('disabled', false)}
      disableClear={boolean('disableClear', false)}
      hideChevron={boolean('hideChevron', false)}
      buttonStyle={select('buttonStyle', buttonStyles, 'bare')}
      displayValue={text('displayValue', 'Select')}
    >
      <Icon icon="px-nav:home" />
    </Dropdown>
  ));
