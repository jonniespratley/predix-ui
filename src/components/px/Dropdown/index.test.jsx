import renderer from 'react-test-renderer';
import React from 'react';
import { shallow, mount } from 'enzyme';
import sinon from 'sinon';
import Dropdown from './';

const items = [
  {"key":"1","val":"iPhone"},
  {"key":"2","val":"Android"},
  {"key":"3","val":"Blackberry"},
  {"key":"4","val":"Windows Phone"},
  {"key":"5","val":"Flip Phone","disabled":true}
];

describe('Dropdown', () => {
  test('should render', () =>{
    const wrapper = shallow(
      <Dropdown/>
    );
    expect(wrapper.find('.px-dropdown')).toHaveLength(1);
  });

  test('matches snapshot', () => {
    const tree = renderer.create(
      <Dropdown
        displayValue='My Val'
        buttonStyle='default'
        items={items} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should render with items', () =>{
    const wrapper = shallow(
      <Dropdown items={items}/>
    );
    expect(wrapper.find('.px-button')).toHaveLength(1);
    expect(wrapper.find('.px-dropdown__option')).toHaveLength(items.length);
  });

  test('should close when close icon clicked', () =>{
    const wrapper = shallow(
      <Dropdown items={items}/>
    );
    wrapper.setState({opened: true});
    //expect(wrapper.find({icon: 'px-utl:chevron'})).toHaveLength(1);
  });

  test('should clear selected when px-utl-close is clicked', () =>{
    const wrapper = mount(
      <Dropdown items={items} selected={1}/>
    );
    expect(wrapper.state().selected).toEqual(1);
    //expect(wrapper.find({icon: 'px-utl:chevron'})).toHaveLength(1);
  });

  test('should trigger onChange when item is changed', () =>{
    let spy = sinon.spy();
    const wrapper = shallow(
      <Dropdown items={items} onChange={spy} />
    );
    expect(wrapper.find('.px-button')).toHaveLength(1);
    expect(wrapper.find('.px-dropdown__option')).toHaveLength(items.length);
    wrapper.find('[data-val="iPhone"]').simulate('click');
    console.log(wrapper.state());
    //console.log(spy.getCalls());
    expect(spy.calledOnce).toEqual(true);
    //expect(spy.calledWith('test')).toEqual(true);
  });
});
