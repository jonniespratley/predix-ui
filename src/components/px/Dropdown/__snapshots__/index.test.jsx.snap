// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown matches snapshot 1`] = `
.c3 {
  color: var(--iron-icon-stroke-color,inherit);
  fill: var(--iron-icon-fill-color,none);
  stroke: var(--iron-icon-stroke-color,currentColor);
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  vertical-align: middle;
  width: var(--iron-icon-width);
  height: var(--iron-icon-height);
  color: inherit !important;
}

.c3 > svg {
  color: var(--iron-icon-stroke-color,inherit);
  fill: var(--iron-icon-fill-color,none);
  stroke: var(--iron-icon-stroke-color,currentColor);
  position: relative;
  vertical-align: middle;
}

.c3.hidden {
  display: none;
}

.c0 {
  font-family: "GE Inspira Sans";
  outline: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: inline-block;
  overflow: visible;
  height: var(--px-btn-height,auto);
  min-height: var(--px-btn-min-height,2rem);
  min-width: var(--px-btn-min-width,4.66667em);
  margin: 0;
  border: 1px solid var(--px-btn-border-color,transparent);
  border-radius: 0 !important;
  padding: calc(var(--px-btn-height,2em) / 10) calc(var(--px-btn-height,2em) / 2);
  box-shadow: var(--px-btn-shadow--light,none);
  font: inherit;
  line-height: calc(var(--px-btn-height,2em) - 2px);
  cursor: pointer;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: none;
  white-space: nowrap;
  background-color: var(--px-btn-background,#d8e0e5);
  -webkit-transition: background .4s,border-color .4s,color .4s;
  transition: background .4s,border-color .4s,color .4s;
  -webkit-appearance: button;
  -webkit-font-smoothing: antialiased;
  color: var(--px-btn-color,#2c404c);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0:link,
.c0:visited,
.c0:hover,
.c0:active {
  color: var(--px-btn-color,#2c404c);
}

.c0:hover,
.c0:focus {
  border-color: var(--px-btn-border-color--hover,transparent);
  box-shadow: var(--px-btn-shadow,none);
  background-color: var(--px-btn-background--hover,#a3b5bf);
}

.c0:active {
  outline: none;
  border-color: var(--px-btn-border-color--pressed,transparent);
  box-shadow: none;
  background-color: var(--px-btn-background--pressed,#889aa5);
}

.c2 {
  overflow: hidden;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  text-overflow: ellipsis;
  text-align: left;
}

.c2 ~ i {
  margin-left: .5rem;
}

.c4 {
  z-index: 10;
  position: absolute;
  background-color: var(--px-dropdown-bg-color,#fff);
  border: 1px solid var(--px-dropdown-border-color,gray);
  max-height: var(--px-dropdown-max-height,300px);
  max-width: var(--px-dropdown-max-width,400px);
  min-width: var(--px-dropdown-min-width);
  min-height: var(--px-dropdown-min-height);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  display: none;
}

.c5 {
  padding-left: var(--px-dropdown-item-padding,.66667rem);
  padding-right: var(--px-dropdown-item-padding,.66667rem);
  padding-top: var(--px-dropdown-item-padding);
  padding-bottom: var(--px-dropdown-item-padding);
  height: var(--px-dropdown-item-height,2rem);
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c5:hover {
  background-color: var(--px-dropdown-bg-color--hover,#d3d3d3);
}

.c5.selected,
.c5:active {
  background-color: var(--px-dropdown-bg-color--selected,#00f);
  color: var(--px-dropdown-text-color--selected,#fff);
}

.c6 {
  padding-left: var(--px-dropdown-item-padding,.66667rem);
  padding-right: var(--px-dropdown-item-padding,.66667rem);
  padding-top: var(--px-dropdown-item-padding);
  padding-bottom: var(--px-dropdown-item-padding);
  height: var(--px-dropdown-item-height,2rem);
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: var(--px-dropdown-text-color--disabled,gray);
  background-color: var(--px-dropdown-bg-color--disabled,transparent);
  cursor: not-allowed;
  pointer-events: none;
}

.c6:hover {
  background-color: var(--px-dropdown-bg-color--hover,#d3d3d3);
}

.c6.selected,
.c6:active {
  background-color: var(--px-dropdown-bg-color--selected,#00f);
  color: var(--px-dropdown-text-color--selected,#fff);
}

.c1 {
  max-width: inherit;
  width: inherit;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  outline: 0;
  font-weight: var(--px-dropdown-font-weight,normal);
  height: auto;
}

<div
  className="px-dropdown"
>
  <button
    className="px-button c0"
    disabled={null}
    onClick={[Function]}
    size={null}
    style={
      Object {
        "height": null,
      }
    }
    type={null}
  >
    <div
      className="px-dropdown__trigger c1"
    >
      <div
        className="px-dropdown__label c2"
      >
        My Val
      </div>
      <i
        className="px-icon px-utl-chevron c3"
        color="inherit"
        style={
          Object {
            "height": 16,
            "width": 16,
          }
        }
      >
        <svg
          dangerouslySetInnerHTML={
            Object {
              "__html": "<g id=\\"px-utl-chevron\\"><path d=\\"M2.4 6.2l5.5 5.5 5.5-5.5\\"/></g>",
            }
          }
          preserveAspectRatio="xMidYMid meet"
          style={
            Object {
              "display": "block",
              "height": 16,
              "pointerEvents": "none",
              "width": 16,
            }
          }
          viewBox="0 0 16 16"
        />
      </i>
    </div>
  </button>
  <div
    className="px-dropdown__content c4"
    width={null}
  >
    <div
      className="px-dropdown__option c5"
      data-key="1"
      data-val="iPhone"
      onClick={[Function]}
    >
      iPhone
    </div>
    <div
      className="px-dropdown__option c5"
      data-key="2"
      data-val="Android"
      onClick={[Function]}
    >
      Android
    </div>
    <div
      className="px-dropdown__option c5"
      data-key="3"
      data-val="Blackberry"
      onClick={[Function]}
    >
      Blackberry
    </div>
    <div
      className="px-dropdown__option c5"
      data-key="4"
      data-val="Windows Phone"
      onClick={[Function]}
    >
      Windows Phone
    </div>
    <div
      className="px-dropdown__option c6"
      data-key="5"
      data-val="Flip Phone"
      disabled={true}
      onClick={[Function]}
    >
      Flip Phone
    </div>
  </div>
</div>
`;
